sui_core/src/gateway_types.rs

```diff
@@ -0,0 +1,947 @@
+    pub storage_rebate: u64,
+    pub reference: SuiObjectRef,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema, Eq, PartialEq, Ord, PartialOrd)]
+#[serde(rename_all = "camelCase", rename = "ObjectRef")]
+pub struct SuiObjectRef {
+    /// Hex code as string representing the object id
+    pub object_id: ObjectID,
+    /// Object version.
+    pub version: SequenceNumber,
+    /// Base64 string representing the object digest
+    pub digest: ObjectDigest,
+}
+
+impl SuiObjectRef {
+    pub fn to_object_ref(&self) -> ObjectRef {
+        (self.object_id, self.version, self.digest)
+    }
+}
+
+impl From<ObjectRef> for SuiObjectRef {
+    fn from(oref: ObjectRef) -> Self {
+        Self {
+            object_id: oref.0,
+            version: oref.1,
+            digest: oref.2,
+        }
+    }
+}
+
+impl Display for SuiObject {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let type_string = self
+            .data
+            .type_()
+            .map_or("Move Package".to_owned(), |type_| type_.to_string());
+
+        write!(
+            f,
+            "ID: {:?}\nVersion: {:?}\nOwner: {}\nType: {}",
+            self.id(),
+            self.version().value(),
+            self.owner,
+            type_string
+        )
+    }
+}
+
+impl SuiObject {
+    pub fn id(&self) -> ObjectID {
+        self.reference.object_id
+    }
+    pub fn version(&self) -> SequenceNumber {
+        self.reference.version
+    }
+
+    pub fn try_from(o: Object, layout: Option<MoveStructLayout>) -> Result<Self, anyhow::Error> {
+        let oref = o.compute_object_reference();
+        let data = match o.data {
+            Data::Move(m) => {
+                let move_struct =
+                    m.to_move_value(&layout.ok_or(SuiError::ObjectSerializationError {
+                        error: "Layout is required to convert Move object to json".to_owned(),
+                    })?)?;
+                SuiData::MoveObject(SuiMoveObject {
+                    type_: m.type_.to_string(),
+                    contents: move_struct.into(),
+                })
+            }
+            Data::Package(p) => SuiData::Package(SuiMovePackage {
+                disassembled: p.disassemble()?,
+            }),
+        };
+        Ok(Self {
+            data,
+            owner: o.owner,
+            previous_transaction: o.previous_transaction,
+            storage_rebate: o.storage_rebate,
+            reference: oref.into(),
+        })
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, Eq, PartialEq)]
+#[serde(tag = "dataType", rename_all = "camelCase", rename = "Data")]
+pub enum SuiData {
+    MoveObject(SuiMoveObject),
+    Package(SuiMovePackage),
+}
+
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, Eq, PartialEq)]
+#[serde(rename = "MoveObject")]
+pub struct SuiMoveObject {
+    #[serde(rename = "type")]
+    pub type_: String,
+    pub contents: SuiMoveValue,
```

Comment:
I feel like the type of content can be more restrictive here. Here we are saying the `contents` can be basically any type(e.g., u8, boolean, etc). But isn't the `contents` of `SuiMoveObject` always a struct-like value?

Reply:
Using SuiMoveValue here can allow us to treat some Sui types (GasCoin) as base type instead of struct... or maybe it's not worth treating Coin special? What's your thought?